keeper/upkeep_executer.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	result, err := executer.ethClient.CallContract(context.Background(), msg, nil)
+	if err != nil {
+		// don't log anything as this is extremely common and would be too noisey
+		// error just signifies "inelligible to performUpkeep"
+		return
+	}
+
+	res, err := upkeepRegistryABI.Unpack(checkUpkeep, result)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	performData, ok := res[0].([]byte)
+	if !ok {
+		logger.Error("checkupkeep payload not as expected")
+		return
+	}
+
+	logger.Debugf("Performing upkeep on registry %s, ID %d", registration.Registry.Address.Hex(), registration.UpkeepID)
+
+	performPayload, err := upkeepRegistryABI.Pack(
+		performUpkeep,
+		big.NewInt(int64(registration.UpkeepID)),
+		performData,
+	)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	performPayloadString := utils.AddHexPrefix(common.Bytes2Hex(performPayload[4:]))
+
+	// TODO - RYAN - need to include gas here
+	// https://www.pivotaltracker.com/story/show/176391267
+	chainlinkPayloadJSON := map[string]interface{}{
+		"format":           "preformatted",
+		"address":          registration.Registry.Address.Hex(),
+		"functionSelector": performUpkeepHex,
+		"result":           performPayloadString,
+		"fromAddresses":    []string{registration.Registry.From.Hex()},
+	}
+
+	chainlinkPayload, err := json.Marshal(chainlinkPayloadJSON)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	err = executer.chainlinkNode.TriggerJob(registration.Registry.JobID.String(), chainlinkPayload)
+	if err != nil {
+		logger.Errorf("Unable to trigger job on chainlink node: %v", err)
+	}
+
+	err = executer.registryStore.SetRanAt(registration, executer.blockHeight.Load())
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+}
+
+func (executer upkeepExecuter) setRunsOnHeadSubscription() {
+	logger.Debug("setting UpkeepExecuter to run on new heads")
+
+	headers := make(chan *types.Header)
+	sub, err := executer.ethClient.SubscribeNewHead(context.Background(), headers)
+	defer sub.Unsubscribe()
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case err := <-sub.Err():
+			logger.Errorf("error in keeper header subscription: %v", err)
+		case head := <-headers:
+			executer.blockHeight.Store(uint64(head.Number.Int64()))
+			executer.signalRun()
+		}
+	}
+}
+
+func (executer upkeepExecuter) setRunsOnTimer() {
+	logger.Debug("setting UpkeepExecuter to run on timer")
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case <-ticker.C:
+			height, err := executer.ethClient.BlockNumber(context.TODO())
+			if err != nil {
+				logger.Errorf("unable to determine latest block height: %v", err)
+				continue
+			}
+			if executer.blockHeight.Load() < height {
+				executer.blockHeight.Store(height)
+				executer.signalRun()
+			}
+		}
+	}
+}
+
+func (executer upkeepExecuter) signalRun() {
+	// avoid blocking if signal already in buffer
```

Comment:
Maybe worth considering a `sync.Cond` here.

Reply:
I think I need just as much glue code to use a `sync.Cond` as to use a buffered channel here